syntax = "proto3";
service Game {
    //FIXME: Instead of streaming whole state server should
    //stream enemy actions and new tiles position. Whole state should be
    //returned only on reconnects
    //TODO: Lobby
 	rpc gameSession  (stream GameSessionRequest)    returns (stream GameSessionResponse) {}
}

enum GameAction {
    ACTION_LEFT_MOVE = 0;
    ACTION_RIGHT_MOVE = 1;
    ACTION_UP_MOVE = 2;
    ACTION_DOWN_MOVE = 3;
}

message GameState {
    repeated int32 tile = 1; //Power of 2
	bool your_turn_now = 2;
    uint64 your_score = 3;
    uint64 enemy_score = 4;
}

enum ActionStatus {
    STATUS_OK = 0;
    STATUS_INVALID_ACTION =1;
}

message ActionResult {
    ActionStatus status = 1;
    GameState resultState = 2;
}
message GameSessionResponse {
    oneof response {
          ActionResult result = 1;
    }
}

message GameSessionRequest {
    oneof request {
          GameAction action = 1;
    }
}
